### PR TITLE
kms: key rotation support for vault kms

### PR DESCRIPTION
**Pull request by iPraveenParihar**
Rook already has Key rotation algorithm implemented for default type https://github.com/rook/rook/pull/11749 (encryption keys are stored in a K8s Secret).
This PR adds key rotation support for vault KMS using the same algorithm.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: Documentation/Storage-Configuration/Advanced/key-management-system.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ The `security` section contains settings related to encryption of the cluster.
         * `schedule`: the schedule, written in [cron format](https://en.wikipedia.org/wiki/Cron), with which key rotation [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is created, default value is `"@weekly"`.
 
 !!! note
-    Currently key rotation is only supported for the default type, where the Key Encryption Keys are stored in a Kubernetes Secret.
+    Currently key rotation is supported for the default type, where the Key Encryption Keys are stored in a Kubernetes Secret and Vault KMS.
```
Comment (sp98) [inline Documentation/Storage-Configuration/Advanced/key-management-system.md:24]:
```suggestion
    Currently key rotation is supported when the Key Encryption Keys are stored in a Kubernetes Secret and Vault KMS.
```

`Vault` is not default type. So I think the text should be changed as above.

**Review by sp98 — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/kms/kms.go [hunk 2]
```diff
@@ -238,6 +238,25 @@ func (c *Config) UpdateSecret(secretName, secretValue string) error {
 
 		return nil
 	}
+	if c.IsVault() {
+		// Store the secret in Vault
+		v, err := InitVault(c.ClusterInfo.Context, c.context, c.ClusterInfo.Namespace, c.clusterSpec.Security.KeyManagementService.ConnectionDetails)
```
Comment (parth-gr) [inline pkg/daemon/ceph/osd/kms/kms.go:243]: can you add this cluster yam CR spec?

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/kms/kms.go (hunk 2 above)
Comment (iPraveenParihar) [reply pkg/daemon/ceph/osd/kms/kms.go:243]: No spec change is needed.  KeyRotation spec was already added [here](https://github.com/rook/rook/commit/5dc8f5a688c72fb95dc032d001b8b7ed47e78cae).

File: Documentation/Storage-Configuration/Advanced/key-management-system.md [hunk 3]
```diff
@@ -21,7 +21,7 @@ The `security` section contains settings related to encryption of the cluster.
         * `schedule`: the schedule, written in [cron format](https://en.wikipedia.org/wiki/Cron), with which key rotation [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is created, default value is `"@weekly"`.
 
 !!! note
-    Currently key rotation is only supported for the default type, where the Key Encryption Keys are stored in a Kubernetes Secret.
+    Currently key rotation is supported when the Key Encryption Keys are stored in a Kubernetes Secret and Vault KMS.
```
Comment (Rakshith-R) [inline Documentation/Storage-Configuration/Advanced/key-management-system.md:24]:
```suggestion
    Currently key rotation is supported when the Key Encryption Keys are stored in Kubernetes Secret or Vault KMS.
```

**Review by Rakshith-R — COMMENTED** [review]
tiny nit
LGTM

**Review by Rakshith-R — APPROVED** [review]
Thanks

File: pkg/daemon/ceph/osd/kms/kms.go [hunk 4]
```diff
@@ -238,6 +238,25 @@ func (c *Config) UpdateSecret(secretName, secretValue string) error {
...
+		if err != nil {
+			return errors.Wrap(err, "failed to init vault kms")
+		}
+		k := buildVaultKeyContext(c.clusterSpec.Security.KeyManagementService.ConnectionDetails)
+		// Build Secret
+		secretName = GenerateOSDEncryptionSecretName(secretName)
+		data := make(map[string]interface{})
```
Comment (Madhu-1) [inline pkg/daemon/ceph/osd/kms/kms.go:250]:
```suggestion
data := make(map[string]string)
```
secretValue is of type string, why not use that as we know the type?

File: tests/scripts/deploy-validate-vault.sh [hunk 5]
```diff
@@ -276,6 +276,34 @@ function validate_osd_secret {
...
+    sleep 10s
+  done
+  new_key=$(kubectl exec vault-0 -- sh -c "$cmd" | jq -r ".data.\"$key_name\"")
+  if [ "$old_key" == "$new_key" ]; then
+    echo "encryption passphrase is not rotated"
+    exit 1
+  else
+    echo "encryption passphrase is successfully rotated"
+  fi
+}
```
Comment (Madhu-1) [inline tests/scripts/deploy-validate-vault.sh:305]:
```suggestion
runtime=180  # 3 minutes in seconds
endtime=$((SECONDS + runtime))
 while [ $SECONDS -le $endtime ]; do
    echo "Time Now: $(date +%H:%M:%S)"
    local new_key=$(kubectl exec vault-0 -- sh -c "$cmd" | jq -r ".data.\"$key_name\"")
    
    if [ "$old_key" != "$new_key" ]; then
      echo "Encryption passphrase successfully rotated."
      exit 0
    fi
    
    echo "Encryption passphrase not rotated, sleeping for 10 seconds."
    sleep 10
  done
  
  echo "Encryption passphrase not rotated."
  exit 1
}
```

can we do like as suggested as we are using seconds to sleep and also this reduces a few complex checks?

File: pkg/daemon/ceph/osd/kms/kms.go [hunk 6]
```diff
@@ -238,6 +238,25 @@ func (c *Config) UpdateSecret(secretName, secretValue string) error {
...
+		data[secretName] = secretValue
+
+		_, err = v.PutSecret(secretName, data, k)
+		if err != nil {
+			return errors.Wrapf(err, "failed to put secret %q in kms", secretName)
```
Comment (Madhu-1) [inline pkg/daemon/ceph/osd/kms/kms.go:255]:
```suggestion
return errors.Wrapf(err, "failed to put secret %q in vault kms", secretName)
```

File: pkg/daemon/ceph/osd/kms/kms.go (hunk 4 above)
Comment (iPraveenParihar) [reply pkg/daemon/ceph/osd/kms/kms.go:250]: Rook uses [libopenstorage/secrets](https://github.com/libopenstorage/secrets) library for KMS and its implementation of PutSecret is as follows
```
PutSecret(
	secretId string,
	plainText map[string]interface{},
	keyContext map[string]string,
) (Version, error)
```

**Review by Madhu-1 — COMMENTED** [review]
LGTM, please check the shell check failure

**Comment by iPraveenParihar** [comment]
> LGTM, please check the shell check failure

Done, thanks!

**Review by sp98 — APPROVED** [review]

**Review by parth-gr — APPROVED** [review]

**Pull request merged**